From 20f903512c7d756bf398eaf4239dff1bc6cdd7df Mon Sep 17 00:00:00 2001
From: Shunsuke Kataoka <shunsuke.kataoka.df@renesas.com>
Date: Wed, 26 Jun 2013 13:36:17 +0900
Subject: [PATCH 672/715] ASoC: sh: scu: Fix SRC initialize process

Modify the interface (argument) of callback function for SRC process.
Move the SRC process that is executed in the PCM function to the callback
function.

Signed-off-by: Shunsuke Kataoka <shunsuke.kataoka.df@renesas.com>
(cherry picked from commit f2c2a9bc969ffbf0356315de85917733b51e2cdc)

Signed-off-by: Ryo Kataoka <ryo.kataoka.wt@renesas.com>
---
 include/sound/sh_scu.h |    9 ++++-----
 sound/soc/sh/scu_dai.c |   40 ++++++++++++++++++++++++++++------------
 sound/soc/sh/scu_pcm.c |   12 ++++--------
 3 files changed, 36 insertions(+), 25 deletions(-)

diff --git a/include/sound/sh_scu.h b/include/sound/sh_scu.h
index 9e6ae85..484074d 100644
--- a/include/sound/sh_scu.h
+++ b/include/sound/sh_scu.h
@@ -1976,7 +1976,7 @@ struct scu_playback_callback {
 	void (*init_ssi)(void);
 	void (*init_ssi_src)(void);
 	void (*init_ssi_dvc)(void);
-	void (*init_src)(void);
+	void (*init_src)(unsigned int);
 	void (*init_dvc)(void);
 	void (*deinit_ssi)(void);
 	void (*deinit_ssi_src)(void);
@@ -1989,7 +1989,7 @@ struct scu_capture_callback {
 	void (*init_ssi)(void);
 	void (*init_ssi_src)(void);
 	void (*init_ssi_dvc)(void);
-	void (*init_src)(void);
+	void (*init_src)(unsigned int);
 	void (*init_src_dvc)(void);
 	void (*init_dvc)(void);
 	void (*deinit_ssi)(void);
@@ -2055,12 +2055,11 @@ struct scu_pcm_info {
 extern struct snd_soc_platform_driver scu_platform;
 
 extern struct scu_route_info *scu_get_route_info(void);
-extern void scu_src_control(int src_ch, struct snd_pcm_substream *ss);
 
 extern void scu_init_ssi0(void);
 extern void scu_init_ssi0_src0(void);
 extern void scu_init_ssi0_dvc0(void);
-extern void scu_init_src0(void);
+extern void scu_init_src0(unsigned int);
 extern void scu_init_dvc0(void);
 extern void scu_deinit_ssi0(void);
 extern void scu_deinit_ssi0_src0(void);
@@ -2071,7 +2070,7 @@ extern void scu_deinit_dvc0(void);
 extern void scu_init_ssi1(void);
 extern void scu_init_ssi1_src1(void);
 extern void scu_init_ssi1_dvc1(void);
-extern void scu_init_src1(void);
+extern void scu_init_src1(unsigned int);
 extern void scu_init_src1_dvc1(void);
 extern void scu_init_dvc1(void);
 extern void scu_deinit_ssi1(void);
diff --git a/sound/soc/sh/scu_dai.c b/sound/soc/sh/scu_dai.c
index e3ddb72..29b1f9a 100644
--- a/sound/soc/sh/scu_dai.c
+++ b/sound/soc/sh/scu_dai.c
@@ -187,7 +187,25 @@ static void scu_ssi_stop(int ssi_ch)
 	writel(0, (u32 *)(rinfo->ssiureg + reg));
 }
 
-void scu_src_control(int src_ch, struct snd_pcm_substream *ss)
+static void scu_src_init(int src_ch)
+{
+	FNC_ENTRY
+	/* SCU SRC_MODE */
+	writel(SRC_MODE_SRCUSE, (u32 *)&rinfo->scusrcreg[src_ch]->mode);
+	FNC_EXIT
+	return;
+}
+
+static void scu_src_deinit(int src_ch)
+{
+	FNC_ENTRY
+	/* SCU SRC_MODE */
+	writel(0, (u32 *)&rinfo->scusrcreg[src_ch]->mode);
+	FNC_EXIT
+	return;
+}
+
+void scu_src_control(int src_ch, unsigned int rate)
 {
 	u64 val = 0;
 
@@ -209,7 +227,7 @@ void scu_src_control(int src_ch, struct snd_pcm_substream *ss)
 	/* SRC_IFSVR INTIFS calculation */
 #ifdef CONVERT_48KHZ
 	/* Convert example (48kHz) */
-	val = div_u64(SRC_IFS_FSO * ss->runtime->rate, SRC_IFS_48KHZ);
+	val = div_u64(SRC_IFS_FSO * rate, SRC_IFS_48KHZ);
 #else
 	/* Not convert example (CODEC driver converts by itself) */
 	val = SRC_IFS_FSO;
@@ -235,14 +253,10 @@ void scu_src_control(int src_ch, struct snd_pcm_substream *ss)
 	FNC_EXIT
 	return;
 }
-EXPORT_SYMBOL(scu_src_control);
 
 static void scu_src_start(int src_ch, int src_dir)
 {
 	FNC_ENTRY
-	/* SCU SRC_MODE */
-	writel(SRC_MODE_SRCUSE, (u32 *)&rinfo->scusrcreg[src_ch]->mode);
-
 	/* SRC_CONTROL */
 	if (src_dir == SRC_INOUT)
 		scu_or_writel((SRC_MODE_START_IN | SRC_MODE_START_OUT),
@@ -270,10 +284,6 @@ static void scu_src_stop(int src_ch, int src_dir)
 	else /* SRC_IN */
 		scu_and_writel(~SRC_MODE_START_IN,
 			(u32 *)&rinfo->scusrcreg[src_ch]->control);
-
-	/* SCU SRC_MODE */
-	writel(0, (u32 *)&rinfo->scusrcreg[src_ch]->mode);
-
 	FNC_EXIT
 	return;
 }
@@ -347,8 +357,10 @@ void scu_deinit_ssi0_dvc0(void)
 }
 EXPORT_SYMBOL(scu_deinit_ssi0_dvc0);
 
-void scu_init_src0(void)
+void scu_init_src0(unsigned int rate)
 {
+	scu_src_init(0);
+	scu_src_control(0, rate);
 	/* start src */
 	scu_src_start(0, SRC_INOUT);
 }
@@ -358,6 +370,7 @@ void scu_deinit_src0(void)
 {
 	/* stop src */
 	scu_src_stop(0, SRC_INOUT);
+	scu_src_deinit(0);
 }
 EXPORT_SYMBOL(scu_deinit_src0);
 
@@ -445,8 +458,10 @@ void scu_deinit_ssi1_dvc1(void)
 }
 EXPORT_SYMBOL(scu_deinit_ssi1_dvc1);
 
-void scu_init_src1(void)
+void scu_init_src1(unsigned int rate)
 {
+	scu_src_init(1);
+	scu_src_control(1, rate);
 	/* start src */
 	scu_src_start(1, SRC_INOUT);
 }
@@ -456,6 +471,7 @@ void scu_deinit_src1(void)
 {
 	/* stop src */
 	scu_src_stop(1, SRC_INOUT);
+	scu_src_deinit(1);
 }
 EXPORT_SYMBOL(scu_deinit_src1);
 
diff --git a/sound/soc/sh/scu_pcm.c b/sound/soc/sh/scu_pcm.c
index f4a436e..d9c2eb8 100644
--- a/sound/soc/sh/scu_pcm.c
+++ b/sound/soc/sh/scu_pcm.c
@@ -376,14 +376,12 @@ static void scu_pcm_start(struct snd_pcm_substream *ss, int first_flag)
 			scu_audma_start(SHDMA_SLAVE_PCM_MEM_SRC0, ss);
 
 			if (first_flag) {
-				/* control src */
-				scu_src_control(0, ss);
-
 				/* start ssi */
 				pcminfo->routeinfo->pcb.init_ssi_src();
 
 				/* start src */
-				pcminfo->routeinfo->pcb.init_src();
+				pcminfo->routeinfo->pcb.init_src(
+					ss->runtime->rate);
 			}
 		}
 
@@ -409,14 +407,12 @@ static void scu_pcm_start(struct snd_pcm_substream *ss, int first_flag)
 			scu_audma_start(SHDMA_SLAVE_PCM_SRC1_MEM, ss);
 
 			if (first_flag) {
-				/* control src */
-				scu_src_control(1, ss);
-
 				/* start ssi */
 				pcminfo->routeinfo->ccb.init_ssi_src();
 
 				/* start src */
-				pcminfo->routeinfo->ccb.init_src();
+				pcminfo->routeinfo->ccb.init_src(
+					ss->runtime->rate);
 			}
 		}
 
-- 
1.7.10.4

